### PR TITLE
RDM-13130 Add CategoryId to CaseField and Validation

### PR DESCRIPTION
**Pull request by CameronGuthrie1**
[###] JIRA link (if applicable) ###
[RDM-13130](https://tools.hmcts.net/jira/browse/RDM-13130) _[Sub-task] CategoryID column in CaseField tab_

[RDM-13079](https://tools.hmcts.net/jira/browse/RDM-13079) _[Parent-task]_ _CategoryID column in CaseField and ComplexTypes tab_ 

### Change description ###
Add CategoryId to the CaseFiled Tab and then validate that the caseField is the correct type and has a categoryId that exists in the categories tab

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java [hunk 1]
```diff
@@ -153,6 +159,31 @@ public ParseContext registerStateForCaseType(String caseTypeId, StateEntity stat
...
+        final Map<String, CategoryEntity> categoryStates = categoryByCaseTypes.get(caseTypeId);
+
+        if (categoryStates.containsKey(categoryEntity.getCategoryId())) {
+            throw new SpreadsheetParsingException("Category already registered for ID: "
+                + categoryEntity.getCategoryId());
+        }
+
+        categoryStates.put(categoryEntity.getCategoryId(), categoryEntity);
+
+        return this;
```
Comment (sullivangareth) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java:184]: Obviously don't need this if void method

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParserTest.java [hunk 2]
```diff
@@ -59,6 +59,7 @@ public void testParse() {
         assertThat(categoryEntities.get(0).getCategoryLabel(), is(CATEGORY_LABEL));
         assertThat(categoryEntities.get(0).getDisplayOrder(), is(DISPLAY_ORDER));
         assertThat(categoryEntities.get(0).getParentCategoryId(), is(NO_PARENT));
+        assertThat(parseContext.getCategory(CASE_TYPE, CATEGORY_ID), is(categoryEntities.get(0)));
```
Comment (sullivangareth) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParserTest.java:62]: use import

`import static org.hamcrest.MatcherAssert.assertThat;`

to get rid of deprecation warnings?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 3]
```diff
@@ -55,7 +56,7 @@ public void shouldRegisterAndGetCaseFieldForCaseType() {
         parseContext.registerCaseFieldForCaseType("caseTypeId", caseField);
 
         final CaseFieldEntity retrieved = parseContext.getCaseFieldForCaseType("caseTypeId", "case field id");
-        assertThat(retrieved, is(caseField));
+        Assert.assertThat(retrieved, is(caseField));
```
Comment (sullivangareth) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:59]: use on demand static import ?

**Review by sullivangareth — APPROVED** [review]

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParser.java [hunk 4]
```diff
@@ -71,6 +71,8 @@ public CategoryEntity createCategoriesEntity(ParseContext parseContext,
         categoriesEntity.setParentCategoryId(definitionDataItem.getString(ColumnName.PARENT_CATEGORY_ID));
         categoriesEntity.setCategoryLabel(definitionDataItem.getString(ColumnName.CATEGORY_LABEL));
         categoriesEntity.setCategoryId(definitionDataItem.getString(ColumnName.CATEGORY_ID));
+        //register Case Type and category
+        parseContext.registerCaseTypeForCategory(caseType, categoriesEntity);
```
Comment (sem-hmcts) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParser.java:75]: This is mutating an input parameter, `parseContext` in this case.  I think this code belongs somewhere else.

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidator.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    public CategoryIdValidator() {
+    }
+
+    private void validate(ParseContext parseContext,
+                          SheetName sheetName,
+                          String caseTypeId,
+                          String categoryId,
+                          FieldTypeEntity fieldType) {
+        if (categoryId != null) {
```
Comment (sem-hmcts) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidator.java:31]: Could unwrap these nested `ifs` for readability.

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParserTest.java (hunk 2 above)
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/CategoryParserTest.java:62]: The assertion of `parseContext.getCategory(...)` looks out of place here.  All the other assertions are checking that `categoryEntities` was parsed correctly.
I think we should test one logic per test

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java (hunk 3 above)
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:59]: Junit 4 assert?  Should be using Junit 5

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 6]
```diff
@@ -66,7 +67,7 @@ public void shouldRegisterAndGetCaseFieldType() {
 
         final FieldTypeEntity retrieved = parseContext.getCaseFieldType("caseTypeId", "fieldId");
 
-        assertThat(retrieved, is(fieldType));
+        Assert.assertThat(retrieved, is(fieldType));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:70]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 7]
```diff
@@ -92,7 +93,7 @@ public void shouldRegisterAndGetStateForCaseType() {
 
         final StateEntity retrieved = parseContext.getStateForCaseType("case type id", "some state id");
 
-        assertThat(retrieved, is(state));
+        Assert.assertThat(retrieved, is(state));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:96]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 8]
```diff
@@ -117,7 +118,7 @@ public void shouldSetAndGetJurisdiction() {
         final JurisdictionEntity j = mock(JurisdictionEntity.class);
         parseContext.setJurisdiction(j);
 
-        assertThat(parseContext.getJurisdiction(), is(j));
+        Assert.assertThat(parseContext.getJurisdiction(), is(j));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:121]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 9]
```diff
@@ -130,8 +131,8 @@ public void shouldRegisterAndGetCaseType() {
         parseContext.registerCaseType(t1);
 
         final Set<CaseTypeEntity> caseTypes = parseContext.getCaseTypes();
-        assertThat(caseTypes.size(), is(2));
-        assertThat(caseTypes, containsInAnyOrder(t1, t2));
+        Assert.assertThat(caseTypes.size(), is(2));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:134]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 10]
```diff
@@ -210,7 +211,7 @@ public void shouldFail_whenCaseFieldNotRegisteredForCaseType() {
     @Test
     public void shouldGetEmpty_whenGetBaseTypeNotExist() {
         final Optional<FieldTypeEntity> type = parseContext.getBaseType("type");
-        assertThat(type, isEmpty());
+        Assert.assertThat(type, isEmpty());
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:214]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 11]
```diff
@@ -221,7 +222,7 @@ public void shouldAddAndGetBaseType() {
 
         parseContext.addBaseTypes(Arrays.asList(t));
         final Optional<FieldTypeEntity> type = parseContext.getBaseType("ngitb");
-        assertThat(type.get(), is(t));
+        Assert.assertThat(type.get(), is(t));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:225]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 12]
```diff
@@ -232,14 +233,14 @@ public void shouldAddToAllTypesAndGet() {
 
         parseContext.addToAllTypes(Arrays.asList(t));
         final Optional<FieldTypeEntity> type = parseContext.getType("ngitb");
-        assertThat(type.get(), is(t));
+        Assert.assertThat(type.get(), is(t));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:236]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 13]
```diff
@@ -232,14 +233,14 @@ public void shouldAddToAllTypesAndGet() {
...
     }
 
     @Test
     public void shouldGetEmpty_whenGetTypeNotExist() {
 
         final Optional<FieldTypeEntity> type = parseContext.getType("ngitb");
-        assertThat(type, isEmpty());
+        Assert.assertThat(type, isEmpty());
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:243]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 14]
```diff
@@ -255,6 +256,38 @@ public void shouldReturnMetadataField_whenGetCaseFieldForCaseTypeIsMetadataField
 
         CaseFieldEntity response = parseContext.getCaseFieldForCaseType("caseTypeId", metadataFieldName);
 
-        assertThat(response, is(metadataField));
+        Assert.assertThat(response, is(metadataField));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:259]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 15]
```diff
@@ -255,6 +256,38 @@ public void shouldReturnMetadataField_whenGetCaseFieldForCaseTypeIsMetadataField
...
+
+    @Test
+    public void testRegisterCaseTypeForCategory() {
+
+        final CategoryEntity category = new CategoryEntity();
+        category.setCategoryId("A");
+
+        parseContext.registerCaseTypeForCategory("caseTypeId", category);
+
+        Assert.assertEquals(parseContext.getCategory("caseTypeId", category.getCategoryId()), category);
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:270]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 16]
```diff
@@ -255,6 +256,38 @@ public void shouldReturnMetadataField_whenGetCaseFieldForCaseTypeIsMetadataField
...
+    }
+
+    @Test(expected = SpreadsheetParsingException.class)
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:273]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java [hunk 17]
```diff
@@ -255,6 +256,38 @@ public void shouldReturnMetadataField_whenGetCaseFieldForCaseTypeIsMetadataField
...
+            parseContext.registerCaseTypeForCategory("caseTypeId", category);
+        } catch (SpreadsheetParsingException ex) {
+            Assertions.assertThat(ex).hasMessage("Category already registered for ID: Category");
+            throw ex;
+        }
+    }
+
+    @Test
+    public void testGetCategoryWhenEmpty() {
+        Assert.assertNull(parseContext.getCategory("caseTypeId", "Category"));
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContextTest.java:291]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
+package uk.gov.hmcts.ccd.definition.store.excel.validation;
+
+import lombok.val;
+import org.junit.Before;
+import org.junit.Test;
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java:5]: Junit 4?

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+import org.mockito.MockitoAnnotations;
+import uk.gov.hmcts.ccd.definition.store.excel.endpoint.exception.InvalidImportException;
+import uk.gov.hmcts.ccd.definition.store.excel.parser.ParseContext;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.CaseFieldEntity;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.CaseTypeEntity;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.CategoryEntity;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.FieldTypeEntity;
+
+public class CategoryIdValidatorTest {
```
Comment (sem-hmcts) [inline excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java:14]: The tests in here don't seem to have any assertions

**Review by sem-hmcts — COMMENTED** [review]
I have a few comments

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java [hunk 20]
```diff
@@ -153,6 +159,31 @@ public ParseContext registerStateForCaseType(String caseTypeId, StateEntity stat
...
+        final Map<String, CategoryEntity> categories = categoryByCaseTypes.get(caseTypeId);
+
+        if (null == categories) {
+            return null;
+        }
+
+        return categories.get(categoryId);
+    }
+
+    public ParseContext registerCaseTypeForCategory(String caseTypeId, CategoryEntity categoryEntity) {
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java:172]: done. But will check

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java (hunk 1 above)
Comment (CameronGuthrie1) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java:184]: done but will check

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+        caseField.setCategoryId(category);
+        return caseField;
+    }
+
+    @Test
+    public void testValidateCategoryNull() {
+        parseContext = buildParseContext("Text", null, null);
+
+        CategoryEntity category1 = new CategoryEntity();
+        category1.setCategoryId("divorceDocs");
```
[the remark this answers is not in the document]
Comment (CameronGuthrie1) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/CategoryIdValidatorTest.java:62]: fixed

**Comment by sullivangareth** [comment]
These changes (plus accompanying FTA tests) are now all present in this PR

https://github.com/hmcts/ccd-definition-store-api/pull/1171

Closing this PR and deleting branch as no longer required

**Pull request closed without merging**